src/github.com/stellar/gateway/listener/payment_listener.go

```diff
@@ -166,40 +176,37 @@ func (pl *PaymentListener) onPayment(payment horizon.PaymentResponse) (err error
 		OperationID: payment.ID,
 		ProcessedAt: pl.now(),
 		PagingToken: payment.PagingToken,
+		Status:      "Processing...",
 	}
 
-	return pl.process(payment, dbPayment)
-}
-
-func (pl *PaymentListener) process(payment horizon.PaymentResponse, dbPayment *entities.ReceivedPayment) (err error) {
-	savePayment := func(payment *entities.ReceivedPayment) (err error) {
-		pl.log.Info(payment.Status)
-		err = pl.entityManager.Persist(payment)
+	err = pl.entityManager.Persist(dbPayment)
+	if err != nil {
 		return
 	}
 
+	dbPayment.Status, err = pl.process(payment)
+
+	pl.log.WithFields(logrus.Fields{"status": dbPayment.Status, "err": err}).Error("Payment processed")
+
+	return pl.entityManager.Persist(dbPayment)
+}
+
+func (pl *PaymentListener) process(payment horizon.PaymentResponse) (status string, err error) {
 	if payment.Type != "payment" && payment.Type != "path_payment" {
-		dbPayment.Status = "Not a payment operation"
-		savePayment(dbPayment)
-		return
+		return "Not a payment operation", nil
 	}
 
 	if payment.To != pl.config.Accounts.ReceivingAccountID {
-		dbPayment.Status = "Operation sent not received"
-		savePayment(dbPayment)
-		return nil
+		return "Operation sent not received", nil
 	}
 
 	if !pl.isAssetAllowed(payment.AssetType, payment.AssetCode, payment.AssetIssuer) {
-		dbPayment.Status = "Asset not allowed"
-		savePayment(dbPayment)
-		return nil
+		return "Asset not allowed", nil
 	}
 
 	err = pl.horizon.LoadMemo(&payment)
 	if err != nil {
-		pl.log.Error("Unable to load transaction memo")
-		return err
+		return "Unable to load transaction memo", err
```

Comment:
I think we should avoid returning valid values for both `status` and `error`.  Idiomatic go expects that if err is not nil, then the result will be invalid.